rfcs/0000-nixpkgs-breaking-change-policy.md

```diff
@@ -0,0 +1,75 @@
+- Avoid breaking the `master` branch as much as possible. The `master` branch is used by the `*-unstable` channels and breaking this branch means that no updates can be pushed. This includes critical security updates as well as regular feature updates.
+
+## Procedure
+
+1. The maintainer will prepare a PR with their intended changes.
+2. The maintainer should test a sample of dependent derivations to ensure that their change is not unnecessarily or unintentionally breaking. (Example: Ensure that a critical output file was not forgotten) Note that sometimes it **is** necessary to break all dependent packages, and the maintainer is not required to avoid this.
+3. The maintainer will get the PR reviewed and approved.
+    1. It is **recommended but not required** to have some maintainers of dependent packages involved in the review to review if the breakage is justified.
+4. The maintainer will contact the maintainers of all dependent, broken packages (herein called sub-maintainers).
+    1. The sub-maintainers will be provided a link to the PR with the breaking changes as well as any context that will help them resolve the breakages.
+    2. The sub-maintainers are expected to update the derivations that they maintain within 7 days.
+        1. These changes should be merged to `master` following the regular nixpkgs PR process if they are backwards compatible. Otherwise they will be merged into the branch of the breaking PR.
+5. The maintainer must provide a grace period of at least 7 days from when the sub-maintainers were notified.
+6. The maintainer must merge `master` into their PR branch.
+7. The maintainer must mark all still-broken packages as [broken](https://nixos.org/manual/nixpkgs/stable/#sec-standard-meta-attributes).
+8. The maintainer can now merge to `master`.
+
+This procedure should not result in a failing package build in `master` at any point.
+
+# Drawbacks
+
+This delays the merge of “core” derivations as the author needs to wait for sub-packages to be tested and possibly fixed up to the 7 day threshold.
+
+# Alternatives
+## Shorter grace period
+
+A shorter grace period allows maintainers to move more quickly but requires maintainers of defendant derivations to jump to action quickly which is not always possible for volunteer driven work. For example if a maintainer is on vacation it can’t be expected that they respond in a couple of days.
+
+## Longer grace period
+
+A longer grace period would give derivation maintainers more time to react to changes in dependencies and provide more time to search for replacement maintainers if the original maintainer has abandoned the derivation. The 7 day number was mostly arbitrary and can easily be changed in the future. A longer grace period would also result in more packages being fixed before the breaking change is shipped which would mean that end-users would experience less breakage on channels.
```

Comment:
In certain cases the package could also be dropped. I feel like this should be mentioned here.

Reply:
We discussed this and there was much disagreement of when to drop vs mark broken. I think that dropping because the maintainer doesn't respond in 7 days is a bit aggressive (they could just be on vacation) so I think the best route is just make it a simple "mark broken" and we can leave when to delete derivations to another policy.

Reply:
If I understand Sandro correctly here, this isn't supposed to be the "default measure", but a valid thing to do if the package is e.g. fairly old.

Reply:
That makes sense to me, but I just wanted to keep it simple. In this policy the action for every package is simply to mark as broken. Then we can have some separate process for removing broken packages. I think trying to work a policy in here for when to delete is just unnecessary complication.

Note also that the intention is to create a tool to add the `broken = true` flags so we don't want to need to teach that tool to make this decision. It would be easiest to create a separate tool to raise PRs for old broken packages that will catch all packages, no matter why they were originally marked as broken.

Reply:
I think removing old, unmaintained packages should be done independently of major updates of packages they depend on. Adding this as a valid option in the process adds extra need for discussion which is hindering if you are aiming to keep a defined schedule.